hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/HoodieFileInputFormatBase.java

```diff
@@ -102,18 +138,98 @@ public final void setConf(Configuration conf) {
     }
     return returns.toArray(new FileStatus[0]);
   }
 
+  @Nonnull
+  private List<FileStatus> listStatusForSnapshotMode(JobConf job,
+                                                     Map<String, HoodieTableMetaClient> tableMetaClientMap,
+                                                     List<Path> snapshotPaths) throws IOException {
+    HoodieLocalEngineContext engineContext = new HoodieLocalEngineContext(job);
+    List<FileStatus> targetFiles = new ArrayList<>();
+
+    TypedProperties props = new TypedProperties(new Properties());
+
+    Map<HoodieTableMetaClient, List<Path>> groupedPaths =
+        HoodieInputFormatUtils.groupSnapshotPathsByMetaClient(tableMetaClientMap.values(), snapshotPaths);
+
+    for (Map.Entry<HoodieTableMetaClient, List<Path>> entry : groupedPaths.entrySet()) {
+      HoodieTableMetaClient tableMetaClient = entry.getKey();
+      List<Path> partitionPaths = entry.getValue();
+
+      // Hive job might specify a max commit instant up to which table's state
+      // should be examined. We simply pass it as query's instant to the file-index
+      Option<String> queryCommitInstant =
+          HoodieHiveUtils.getMaxCommit(job, tableMetaClient.getTableConfig().getTableName());
+
+      boolean shouldIncludePendingCommits =
+          HoodieHiveUtils.shouldIncludePendingCommits(job, tableMetaClient.getTableConfig().getTableName());
+
+      HiveHoodieTableFileIndex fileIndex =
+          new HiveHoodieTableFileIndex(
+              engineContext,
+              tableMetaClient,
+              props,
+              HoodieTableQueryType.QUERY_TYPE_SNAPSHOT,
+              partitionPaths,
+              queryCommitInstant,
+              shouldIncludePendingCommits);
+
+      Map<String, Seq<FileSlice>> partitionedFileSlices =
+          JavaConverters.mapAsJavaMapConverter(fileIndex.listFileSlices()).asJava();
+
+      targetFiles.addAll(
+          partitionedFileSlices.values()
+              .stream()
+              .flatMap(seq -> JavaConverters.seqAsJavaListConverter(seq).asJava().stream())
+              .map(fileSlice -> {
+                Option<HoodieBaseFile> baseFileOpt = fileSlice.getBaseFile();
+                Option<HoodieLogFile> latestLogFileOpt = fileSlice.getLatestLogFile();
+                if (baseFileOpt.isPresent()) {
+                  return getFileStatusUnchecked(baseFileOpt);
+                } else if (includeLogFilesForSnapShotView() && latestLogFileOpt.isPresent()) {
+                  return createRealtimeFileStatusUnchecked(latestLogFileOpt.get(), fileSlice.getLogFiles());
+                } else {
+                  throw new IllegalStateException("Invalid state: either base-file or log-file should be present");
+                }
+              })
+              .collect(Collectors.toList())
+      );
+    }
+
+    // TODO cleanup
+    validate(targetFiles, listStatusForSnapshotModeLegacy(job, tableMetaClientMap, snapshotPaths));
```

Comment:
Have a flag to turn off validation by default for query performance?

Reply:
Got it.  I'll take another look once done.

Reply:
Sorry, wasn't elaborate enough -- i'm using this hook to validate the behavior is the same as it was prior to refactoring. I'm planning to clean this up in the topmost PR on the stack.